spot-vaults/contracts/BillBroker.sol

```diff
@@ -775,122 +806,90 @@ contract BillBroker is
                 )
             )
         );
-        if (totalFeePerc >= ONE) {
-            return (0, 0, 0);
+        if (feeFactor >= ONE) {
+            protocolFeePerpAmt = perpAmtOut
+                .mulDiv(feeFactor - ONE, ONE, MathUpgradeable.Rounding.Up)
+                .mulDiv(fees.protocolSwapSharePerc, ONE, MathUpgradeable.Rounding.Up);
         }
-        uint256 totalFeePerpAmt = perpAmtOut.mulDiv(totalFeePerc, ONE);
-        perpAmtOut -= totalFeePerpAmt;
-        protocolFeePerpAmt = totalFeePerpAmt.mulDiv(fees.protocolSwapSharePerc, ONE);
-        lpFeePerpAmt = totalFeePerpAmt - protocolFeePerpAmt;
+        perpAmtOut = perpAmtOut.mulDiv(TWO - feeFactor, ONE);
     }
 
-    /// @notice Computes the swap fee percentage when swapping from perp to usd tokens.
+    /// @notice Computes the swap fee factor when swapping from perp to usd tokens.
     /// @dev Swapping from perp to usd tokens, leaves the system with more perp and fewer usd tokens
     ///      thereby decreasing the system's `assetRatio`. Thus arPost < arPre.
     /// @param arPre The asset ratio of the system before swapping.
     /// @param arPost The asset ratio of the system after swapping.
-    /// @return The fee percentage.
-    function computePerpToUSDSwapFeePerc(
+    /// @return The fee factor.
+    function computePerpToUSDSwapFeeFactor(
         uint256 arPre,
         uint256 arPost
     ) public view returns (uint256) {
-        if (arPost > arPre) {
-            revert UnexpectedARDelta();
+        // When the ar decreases below the lower bound swaps are halted.
+        if (arPost < arHardBound.lower) {
+            return MAX_FEE_FACTOR;
         }
 
-        // When the ar decreases below the lower bound,
-        // swaps are effectively halted by setting fees to 100%.
-        if (arPost < arHardBound.lower) {
-            return ONE;
-        }
-        // When the ar is between the soft and hard bound, a linear function is applied.
-        // When the ar is above the soft bound, a flat percentage fee is applied.
-        //
-        //   fee
-        //    ^
-        //    |
-        // fh |    \          |
-        //    |     \         |
-        //    |      \        |
-        //    |       \       |
-        //    |        \      |
-        //    |         \     |
-        // fl |          \__________
-        //    |               |
-        //    |               |
-        //    |               |
-        //    +---------------------------> ar
-        //       arHL  arSL  1.0
-        //
-        Range memory swapFeePercs = fees.perpToUSDSwapFeePercs;
+        Range memory f1 = fees.perpToUSDSwapFeeFactors;
```

Comment:
Can we rename f1 and f2 to have a reference to perpToUSD or USDToPerp somehow?

Can be abbreviated, like feeP2U or feeU2P